### PR TITLE
Add the BaseEntity mongo document.

### PR DESCRIPTION
**Pull request by frankban**
When adding charms or bundles, always create
the correspnding base entity.

### CODE REVIEW

File: internal/charmstore/store.go [hunk 1]
```diff
@@ -230,6 +233,19 @@ func (s *Store) AddCharm(c charm.Charm, p AddParams) (err error) {
...
+	// Add the base entity to the database.
+	baseEntity := &mongodoc.BaseEntity{
+		URL:    entity.BaseURL,
+		User:   entity.User,
+		Name:   entity.Name,
+		Public: true,
+	}
+	err = s.DB.BaseEntities().Insert(baseEntity)
+	if err != nil && !mgo.IsDup(err) {
+		return errgo.Mask(err)
```
Comment (urosj) [inline internal/charmstore/store.go]: Maybe add annotation like "BaseEntity already exists"?

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/685/
Test PASSed.

File: internal/charmstore/store.go [hunk 2]
```diff
@@ -230,6 +233,19 @@ func (s *Store) AddCharm(c charm.Charm, p AddParams) (err error) {
 }
 
 func (s *Store) insertEntity(entity *mongodoc.Entity) (err error) {
+	// Add the base entity to the database.
+	baseEntity := &mongodoc.BaseEntity{
+		URL:    entity.BaseURL,
+		User:   entity.User,
+		Name:   entity.Name,
+		Public: true,
```
Comment (rogpeppe) [inline internal/charmstore/store.go:242]: Perhaps add a TODO to allow specifying non-public charms on initial upload?

File: internal/charmstore/store_test.go [hunk 3]
```diff
@@ -185,12 +188,27 @@ func (s *StoreSuite) checkAddBundle(c *gc.C, bundle charm.Bundle, addToES bool)
...
+	// Check that the base entity has been properly created.
+	checkBaseEntity(c, store, baseURL(url))
+
 	// Try inserting the bundle again - it should fail because the bundle is
 	// already there.
 	err = store.AddBundleWithArchive(url, bundle)
 	c.Assert(errgo.Cause(err), gc.Equals, params.ErrDuplicateUpload)
 }
 
+func checkBaseEntity(c *gc.C, store *Store, url *charm.Reference) {
```
Comment (rogpeppe) [inline internal/charmstore/store_test.go]: s/check/assert/ ?

**Comment by rogpeppe** [comment]
LGTM modulo a couple of trivials, thanks!

**Comment by urosj** [comment]
:+1:

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/686/
Test PASSed.

**Comment by frankban** [comment]
Thanks for the reviews!
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**